crates/volta-core/src/tool/mod.rs

```diff
@@ -1,660 +1,188 @@
-//! Traits and types for executing command-line tools.
+use std::fmt::{self, Display};
 
-use std::cmp::Ordering;
-use std::env::{self, args_os, ArgsOs};
-use std::ffi::{OsStr, OsString};
-use std::fmt::{self, Debug, Display, Formatter};
-use std::path::Path;
-use std::process::{Command, ExitStatus};
-use std::str::FromStr;
-
-use lazy_static::lazy_static;
-use regex::Regex;
-
-use validate_npm_package_name::{validate, Validity};
-use volta_fail::{throw, Fallible, ResultExt};
-
-use crate::command::create_command;
-use crate::env::UNSAFE_GLOBAL;
 use crate::error::ErrorDetails;
-use crate::path;
-use crate::platform::System;
 use crate::session::Session;
-use crate::signal::pass_control_to_shim;
+use crate::style::{success_prefix, tool_version};
 use crate::version::VersionSpec;
+use log::{debug, info};
+use semver::Version;
+use volta_fail::Fallible;
 
-mod binary;
 mod node;
 mod npm;
-mod npx;
+mod package;
+mod serial;
 mod yarn;
 
-lazy_static! {
-    static ref TOOL_SPEC_PATTERN: Regex =
-        Regex::new("^(?P<name>(?:@([^/]+?)[/])?([^/]+?))(@(?P<version>.+))?$")
-            .expect("regex is valid");
-    static ref HAS_VERSION: Regex = Regex::new(r"^[^\s]+@").expect("regex is valid");
+pub use node::{load_default_npm_version, Node, NodeVersion};
+pub use npm::Npm;
+pub use package::{bin_full_path, BinConfig, BinLoader, Package, PackageDetails};
+pub use yarn::Yarn;
+
+#[inline]
+fn debug_already_fetched<T: Display + Sized>(tool: T) {
+    debug!("{} has already been fetched, skipping download", tool);
+}
+
+#[inline]
+fn info_installed<T: Display + Sized>(tool: T) {
+    info!("{} installed and set {} as default", success_prefix(), tool);
+}
+
+#[inline]
+fn info_fetched<T: Display + Sized>(tool: T) {
+    info!("{} fetched {}", success_prefix(), tool);
 }
 
-/// Distinguish global `add` commands in npm or yarn from all others.
-enum CommandArg {
-    /// The command is a *global* add command.
-    GlobalAdd(Option<OsString>),
-    /// The command is a local, i.e. non-global, add command.
-    NotGlobalAdd,
+#[inline]
+fn info_pinned<T: Display + Sized>(tool: T) {
+    info!("{} pinned {} in package.json", success_prefix(), tool);
+}
+
+/// Trait representing all of the actions that can be taken with a tool
+pub trait Tool: Display {
+    /// Fetch a Tool into the local inventory
+    fn fetch(self, session: &mut Session) -> Fallible<()>;
+    /// Install a tool, making it the default so it is available everywhere on the user's machine
+    fn install(self, session: &mut Session) -> Fallible<()>;
+    /// Pin a tool in the local project so that it is usable within the project
+    fn pin(self, session: &mut Session) -> Fallible<()>;
 }
 
 /// Specification for a tool and its associated version.
-///
-/// Since [`Ord`] is implemented for `ToolSpec`, we can use `.sort` on any
```

Comment:
Oh hello, a prime example of comments getting out of date with code. 😬 Glad this is gone now.